pkg/services/authz/rbac/service.go

```diff
@@ -94,39 +95,84 @@ func (s *Service) Check(ctx context.Context, req *authzv1.CheckRequest) (*authzv
 	return &authzv1.CheckResponse{Allowed: allowed}, nil
 }
 
-func (s *Service) validateRequest(ctx context.Context, req *authzv1.CheckRequest) (*CheckRequest, error) {
+func (s *Service) List(ctx context.Context, req *authzextv1.ListRequest) (*authzextv1.ListResponse, error) {
+	ctx, span := s.tracer.Start(ctx, "authz_direct_db.List")
+	defer span.End()
 	ctxLogger := s.logger.FromContext(ctx)
 
-	if req.GetNamespace() == "" {
-		return nil, status.Error(codes.InvalidArgument, "namespace is required")
+	listReq, err := s.validateListRequest(ctx, req)
+	if err != nil {
+		ctxLogger.Error("invalid request", "error", err)
+		return &authzextv1.ListResponse{}, err
 	}
-	authInfo, has := claims.From(ctx)
-	if !has {
-		return nil, status.Error(codes.Internal, "could not get auth info from context")
+	ctx = request.WithNamespace(ctx, req.GetNamespace())
+
+	permissions, err := s.getUserPermissions(ctx, listReq.Namespace, listReq.UserUID, listReq.Action)
+	if err != nil {
+		ctxLogger.Error("could not get user permissions", "subject", req.GetSubject(), "error", err)
+		return nil, err
+	}
+
+	if permissions["*"] {
+		return &authzextv1.ListResponse{All: true}, nil
+	}
+
+	folderMap, err := s.buildFolderTree(ctx, listReq.Namespace)
+	if err != nil {
+		ctxLogger.Error("could not build folder and dashboard tree", "error", err)
+		return nil, err
+	}
+
+	folderSet := make(map[string]struct{}, len(permissions))
+	dashSet := make(map[string]struct{}, len(permissions))
+	for perm := range permissions {
+		if strings.HasPrefix(perm, "folders:uid:") {
+			identifier := perm[len("folders:uid:"):]
+			folderSet[identifier] = struct{}{}
```

Comment:
We can probably skip already handled folders here:
```suggestion
			if _, ok := folderSet[identifier]; ok {
			 			 continue
			}
			folderSet[identifier] = struct{}{}
```